materialize-snowflake/pipe.go

```diff
@@ -0,0 +1,311 @@
+		return nil, err
+	}
+
+	jwtToken, expiry, err := generateJWTToken(key, cfg.Credentials.User, accountName)
+	if err != nil {
+		return nil, fmt.Errorf("creating jwt token: %w", err)
+	}
+
+	var insertFilesTpl = template.Must(template.New("insertFiles").Parse(insertFilesRawTpl))
+	var insertReportTpl = template.Must(template.New("insertReport").Parse(insertReportRawTpl))
+
+	return &PipeClient{
+		cfg:             cfg,
+		accountName:     accountName,
+		base:            dsnURL.Hostname(),
+		httpClient:      httpClient,
+		token:           jwtToken,
+		expiry:          expiry,
+		insertFilesTpl:  insertFilesTpl,
+		insertReportTpl: insertReportTpl,
+	}, nil
+}
+
+type insertFilesURLTemplate struct {
+	Base     string
+	PipeName string
+	// Used to track the request, we generate a random uuid for this
+	RequestId string
+}
+
+type insertFilesRequest struct {
+	Files []FileRequest `json:"files"`
+}
+
+type FileRequest struct {
+	Path string `json:"path"`
+	Size int    `json:"size"`
+}
+
+type InsertFilesResponse struct {
+	Status    string `json:"responseCode"`
+	RequestId string `json:"requestId"`
+}
+
+const insertFilesRawTpl = "https://{{ $.Base }}/v1/data/pipes/{{ $.PipeName }}/insertFiles?requestId={{ $.RequestId }}"
+const contentType = "application/json;charset=UTF-8"
+const userAgent = "Estuary Technologies Flow"
+
+func (c *PipeClient) refreshJWT() error {
+	if time.Until(c.expiry).Minutes() < 5 {
+		var key, err = c.cfg.Credentials.privateKey()
+		if err != nil {
+			return err
+		}
+		jwtToken, expiry, err := generateJWTToken(key, c.cfg.Credentials.User, c.accountName)
+		if err != nil {
+			return fmt.Errorf("recreating jwt token: %w", err)
+		}
+
+		c.token = jwtToken
+		c.expiry = expiry
+	}
+
+	return nil
+}
+
+// pipeName must be a fully-qualified name, e.g.: database.schema.pipe
+func (c *PipeClient) InsertFiles(pipeName string, files []FileRequest) (*InsertFilesResponse, error) {
+	if err := c.refreshJWT(); err != nil {
+		return nil, err
+	}
+
+	var urlTemplate = insertFilesURLTemplate{
+		Base:      c.base,
+		PipeName:  strings.ToLower(pipeName),
+		RequestId: uuid.New().String(),
+	}
+
+	var reqBody = insertFilesRequest{
+		Files: files,
+	}
+
+	var w strings.Builder
+	if err := c.insertFilesTpl.Execute(&w, urlTemplate); err != nil {
+		return nil, fmt.Errorf("insertFiles template: %w", err)
+	}
+	var url = w.String()
+
+	reqBodyJson, err := json.Marshal(reqBody)
+	if err != nil {
+		return nil, fmt.Errorf("marshal insertFiles body: %w", err)
+	}
+
+	req, err := http.NewRequest("POST", url, bytes.NewReader(reqBodyJson))
+	if err != nil {
+		return nil, fmt.Errorf("creating insertFiles request: %w", err)
+	}
+
+	req.Header.Add("Authorization", fmt.Sprintf("BEARER %s", c.token))
+	req.Header.Add("Content-Type", contentType)
+	req.Header.Add("User-Agent", userAgent)
+
+	resp, err := c.httpClient.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("insertFiles request: %w", err)
+	}
+
+	var respBuf = new(strings.Builder)
```

Comment:
Decoding the response here could be simplified using something like this:

```go
	resp, err := c.httpClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("insertFiles request: %w", err)
	}

	var response InsertFilesResponse
	if err := json.NewDecoder(resp.Body).Decode(&response); err != nil {
		return nil, fmt.Errorf("parsing response of insertFiles failed: %w", err)
	}
	resp.Body.Close()
```

Reply:
I intentionally first write to a string buffer and then decode, since in case of non-200 status code we use the response body as part of the error message.